### ISSUE TITLE
Replace ?showHomeScreen=false with ?bypassHomeScreen

### ISSUE THREAD

**Issue by samreid** [phetsims/joist#451]
In https://github.com/phetsims/joist/issues/445 we are building out more joist functionality regarding the homescreen, and will be adding a new query parameter `?homeScreen=false` which creates a sim with no home screen and no homescreen icon.  There is an existing query parameter `showHomeScreen` which seems like it would be easy to confuse with the new one.  We recommend replacing `?showHomeScreen=false` with `?bypassHomeScreen` for clarity.  We think this is rarely used outside of PhET (probably rarely used within PhET as well), but wanted to check with other developers before making this change.

@jbphet @pixelzoom @jonathanolson @jessegreenberg @Denz1994 @mbarlow12 ok to proceed?  If you are the last one to chime in, please reassign to me or @zepumph.

**Comment by samreid** [comment]
To be clear "bypassHomeScreen" still creates a homescreen, but launches the sim on the 1st screen.  You can get to the homescreen with the navbar home icon.

**Comment by pixelzoom** [comment]
Wouldn't it be more general/flexible to specify which screen to start on?  E.g. `initialScreen=1`?

**Comment by samreid** [comment]
There is already a query parameter `?screenIndex` in which you can specify the initial screen.  We could rename it if desired.

**Comment by pixelzoom** [comment]
`?screenIndex` is the screen that's initially selected on the home screen. It does not start the sim on that screen.

**Comment by pixelzoom** [comment]
(and I'm wonder what the point of it is).

**Comment by pixelzoom** [comment]
Back to `initialScreen`...  Semantics of the values should be the same as `screens` query parameter.  E.g. `?screens=1,3&initialScreen=3`.

**Comment by samreid** [comment]
Currently, you can combine them like so: ?screenIndex=3&showHomeScreen=false

**Comment by pixelzoom** [comment]
`screenIndex` value  apparently has different semantics that `screens`.  E.g. `&screens=1,3&screenIndex=3&showHomeScreen=false` fails.  

If the purpose of `?screenIndex=3&showHomeScreen=false` is to start the sim on a specific screen, why not collapse that to 1 query parameter, like `initialScreen`.

**Comment by pixelzoom** [comment]
I recommend 3 mutually exclusive query parameters:

`screens` - the screens to include, {number[]}
`startScreen` - the screen to display at startup, {number}
`homeScreen` - whether to include the home screen, {boolean}

**Comment by pixelzoom** [comment]
I recommend 3 mutually exclusive query parameters, where all screen numbers refer to indices of the default set of screens, numbered starting from 1.

`screens` - the screens to include, {number[]}
`initialScreen` - the screen to display at startup, {number}
`homeScreen` - whether to include the home screen, {boolean}

Delete `screenIndex` and `showHomeScreen`.

**Comment by pixelzoom** [comment]
At the risk of muddying the waters, I'll propose another alternative that only requires 2 query parameters, `screens` and `initialScreen`.  

This approach identifies the home screen with index `0`. That index would be valid in both `screens` and `initialScreen`, making `homeScreen` unnecessary.  Examples:

For a sim that has 3 non-home screens, the default for `screens` would be `0,1,2,3`.  If you want to exclude the home screen,  use `?screens=1,2,3`.  If you want to include the home screen, but exclude the first non-home screen, use `?screens=0,2,3`. 

The default for `initialScreen` is `0`, the home screen.  If you want to start with the first non-home screen, use `?initialScreen=1`.

The advantage of this approach is 1 less query parameter, and the home screen is treated as a first-class screen. The disadvantage of this approach is that the user has to specify index `0` in `screens`, or they may inadvertently leave out the home screen.  But presumably someone using `screens` should know what they're doing, and will notice that the home screen is missing.

**Comment by zepumph** [comment]
As a developer, I prefer this option. I want to hear from someone like @ariel-phet if the `?screens` is used in the classroom, and how extensively. Is this a large api change for phet? Or are only power users that can adapt easily using this?

**Comment by ariel-phet** [comment]
@zepumph @pixelzoom a few comments:

1. Currently, we have not advertised the ability to select screens to our users, so unlikely that any normal user (or even power user) would know how to set the screens.

2. Although I like the cleanliness of the proposal by @pixelzoom that identifies the home screen with index `0`, my designer side is a bit uncomfortable with the idea. The homescreen serves an entirely different purpose than the other screens in the sim. In addition, we have considered the possibility of using the homescreen in the future for additional settings or such that we do not feel will fit on the navbar. We have not yet made any such expansion, but I just see the homescreen as a different animal, that should be kept cognitively separate from the other screens.

**Comment by pixelzoom** [comment]
Also note that these query parameters can't be changed from their defaults for single-screen sims. I.e.:

`screens:1`, no other values allowed
`initialScreen:1`, no other values allowed
`homeScreen:false`, `true` not allowed

Rather than have to validate that the default haven't been changed for single-screen sims, perhaps it might be easiest to just disallow the use of these options for single-screen sims.

**Comment by pixelzoom** [comment]
@ariel-phet's second comment in https://github.com/phetsims/joist/issues/451#issuecomment-340127235 is unrelated to which approach we use. Both approaches (and this issue) deal with providing the ability to hide the home screen, albeit in slightly different ways.  If we start putting non-navigation controls on the home screen, then (a) we can't hide the home screen, and (b) we need a home screen for single-screen sims (and you never know when a screen that has additional controls might be pulled out into a single-screen sim). 

Imo, "additional settings" should not be solely accessible from the home screen.  They should be accessible from the home screen and the navigation bar, by clicking on some UI control that displays the additional settings.  Whether that's via the PhET menu or an addition control (button with a [typical settings icon](http://fontawesome.io/icon/cog/)) is a design decision.  

So... I don't think that we can provide the ability to hide the home screen without agreement that "additional settings" will not be tied solely to the home screen.

**Comment by samreid** [comment]
https://github.com/phetsims/joist/issues/451#issuecomment-340081541 sounds clearer to me than https://github.com/phetsims/joist/issues/451#issuecomment-340112627

**Comment by zepumph** [comment]
I agree, what do other devs think about proceeding with the following:

>I recommend 3 mutually-exclusive query parameters, where all screen numbers refer to indices of the default set of screens, numbered starting from 1.

>`screens` - the screens to include, {number[]}, defaults to all screens
`initialScreen` - the screen to display at startup, {number}, defaults to home screen
`homeScreen` - whether to include the home screen, {boolean}, defaults to true

>Delete `screenIndex` and `showHomeScreen`.

**Comment by pixelzoom** [comment]
@zepumph asked:
> I agree, what do other devs think about proceeding with the following: ...

_If_ we provide this feature, then that's fine with me.  But as I said in https://github.com/phetsims/joist/issues/451#issuecomment-340202681:
>  I don't think that we can provide the ability to hide the home screen without agreement that "additional settings" will not be tied solely to the home screen.

**Comment by jonathanolson** [comment]
I agree with @pixelzoom's proposals and thoughts above.

**Comment by jbphet** [comment]
I'm in agreement that @pixelzoom's first proposal (https://github.com/phetsims/joist/issues/451#issuecomment-340081541) sounds like a good plan, and I concur that if skipping the home screen is allowed then having "additional settings" on only that screen would be problematic and should be avoided.

**Comment by jbphet** [comment]
FYI, @zepumph said:

> As a developer, I prefer this option. I want to hear from someone like @ariel-phet if the ?screens is used in the classroom, and how extensively. Is this a large api change for phet? Or are only power users that can adapt easily using this?

I just spoke with @kathy-phet about the usage of the `screens` query param (in relation to a different issue), and she told me that it is used by LoL, Nearpod, and by some teachers, so we should be careful not to alter how it works.

**Comment by pixelzoom** [comment]
> ... so we should be careful not to alter how it works.

That's unfortunate, because the way it works now is confusing and not a very good public interface.

**Comment by kathy-phet** [comment]
This still works: 

> I recommend 3 mutually-exclusive query parameters, where all screen numbers refer to indices of the default set of screens, numbered starting from 1.
> screens - the screens to include, {number[]}, defaults to all screens
> initialScreen - the screen to display at startup, {number}, defaults to home screen
> homeScreen - whether to include the home screen, {boolean}, defaults to true
> Delete screenIndex and showHomeScreen.

**Comment by kathy-phet** [comment]
The people I know only use the "screens=x,x" part.

**Comment by pixelzoom** [comment]
It still works.  But if you want consistency across sims (as was requested when we made a similar change to the `screens` query parameter), then that will require republishing sims.

**Comment by pixelzoom** [comment]
Perhaps I misunderstood.  If you're referring only to the `screens` query parameter, then we're likely OK.

**Comment by samreid** [comment]
Rather than adding a new query parameter `initialScreen` and deleting the existing query parameter `screenIndex` we would more likely rename the existing query parameter `screenIndex` to `initialScreen`

**Comment by pixelzoom** [comment]
> ... we would more likely rename the existing query parameter `screenIndex` to `initialScreen`

To clarify... And you proposing to also keep the current behavior, where `initialScreen` would be the initial screen selected on the home screen, and `?initialScreen=3&showHomeScreen=false` would be the required to actually start with that screen visible?  That makes no sense to me - what is the point of using `?initialScreen=3` by itself? And if it's not useful by itself....? 

But if you're going to do that, then please don't use `initialScreen` because it's not going to be the initial screen that you see.  Use something like `selectedScreen` to indicate that it's selected.

**Comment by samreid** [comment]
> Are you proposing to also keep the current behavior, where initialScreen would be the initial screen selected on the home screen

My apologies and you are right, I overlooked the change that initialScreen would bypass the homescreen, so it's not as simple as a rename.

**Comment by zepumph** [comment]
@ariel-phet and I discussed this in the context of "additional settings". We are good to proceed with a HomeScreen query parameter. The current consensus comes from @pixelzoom:
>I recommend 3 mutually-exclusive query parameters, where all screen numbers refer to indices of the default set of screens, numbered starting from 1.
screens - the screens to include, {number[]}, defaults to all screens
initialScreen - the screen to display at startup, {number}, defaults to home screen
homeScreen - whether to include the home screen, {boolean}, defaults to true
Delete screenIndex and showHomeScreen.

Any objections?

**Comment by samreid** [comment]
It sounds correct to me.  What will the default value of the initialScreen be (that indicates to start on the home screen)?

**Comment by pixelzoom** [comment]
```js
   /** 
    * Indicates whether to exclude the home screen for multi-screen sims. 
    * For multi-screen sims only, ignored for single-screen sims (which have no home screen).
    */
    homeScreen: {
        type: 'boolean',
        defaultValue: true
    },

   /** 
    * Specifies the initial screen that will be visible when the sim starts. 
    * See screens query parameter for screen numbering.  The value is one of the values in the 
    * screens array, not an index into the screens value.  For example ?screens=1,3&initialIndex=3,
    * not ?screens=1,3&initialIndex=2.
    * For multi-screen sims only, ignored for single-screen sims.
    */
    initialScreen: {
      type: 'number',
      defautValue: 0 // the home screen
    },

    /**
     * Specifies the set of screens that appear in the sim, and their order.  
     * Uses 1-based numbering (not zero-based) and "," delimited string such as "1,3,4" to get 
     * the 1st, 3rd and 4th screen.  Note that the home screen (screen "0") is included by default. 
     * If you want to exclude the home screen, see the homeScreen query parameter.
     */
    screens: {
      type: 'array',
      elementSchema: {
        type: 'number'
      },
      defaultValue: null
    },
```

**Comment by mbarlow12** [comment]
A quick thought on semantics (from https://github.com/phetsims/joist/issues/451#issuecomment-340817940) and @ariel-phet's design consideration (https://github.com/phetsims/joist/issues/451#issuecomment-340127235). Perhaps `selectedScreen` would be clearer than initialScreen. That would eliminate some potential for troublesome combinations (e.g. `initialScreen:0, homeScreen: false`...that might be the only example, though).

**Comment by pixelzoom** [comment]
@mbarlow12 said:
> Perhaps `selectedScreen` would be clearer than `initialScreen`.

A screen can be selected on the home screen, but not yet visible - which was the problem with the current `screenIndex` query parameter.

**Comment by zepumph** [comment]
>`selectedScreen` 

To be clear, this query parameter is just for startup, there is a `Sim.screenIndexProperty` to keep track of the indices as a user interacts with the sim. Here `initial` seems explicit that it will only effect the startup condition. 

>(e.g. `initialScreen:0`, `homeScreen: false`...that might be the only example, though). The default value of 0 for initial/selected screen would not refer to the home screen in this case.

My understanding is that all indices are based on the full list, so `initialScreen:0`, `homeScreen: false` would throw an error because there is no homeScreen. Am I wrong here? That is how we handle the `screens` qp currently.

**Comment by pixelzoom** [comment]
If you want to handle invalid values for all cases (single- and multi-screen), you'll be looking at a lot of logic.   Another option is to ignore all 3 query parameters for single-screen sims (which I indicated in [the doc above](https://github.com/phetsims/joist/issues/451#issuecomment-340848683)).  Then you could either ignore `initialScreen` unless it's something other than `0', or handle the `initialScreen:0, homeScreen: false` case.

**Comment by pixelzoom** [comment]
Another option is to error out if any of these 3 query parameters are used with a single-screen sim.

**Comment by pixelzoom** [comment]
Another option is to error out (rather than ignore) if any of these 3 query parameters are used with a single-screen sim.  Even if the values are valid, it makes no sense to use these query parameters for a single-screen sim, and even less sense to spend time writing code to support it.

**Comment by zepumph** [comment]
Erroring out on a single screen sim seems reasonable to me as long as the error message is clear that the qp is superfluous/unneeded. 

In general Erroring out may be a better use case for developers so that they can understand how to use these parameters. In this case I think of our LOL clients as developers also, but a teacher may not know how to open the console. . .

**Comment by samreid** [comment]
> The value is one of the values in the screens array, not an index into the screens array. 

It has been requested that we make it possible to clone/modify/shuffle screens for PhET-iO, see https://github.com/phetsims/phet-io/issues/1247.  That will likely introduce a preprocessing step before we attain the screen list.

**Comment by pixelzoom** [comment]
@samreid said
> It has been requested that we make it possible to clone/modify/shuffle screens for PhET-iO, see phetsims/phet-io#1247. That will likely introduce a preprocessing step before we attain the screen list.

Crucial requirement that was missing from this discussion. I'm no longer confident about the recommendations that I've made.

**Comment by kathy-phet** [comment]
For the PhET-iO functionality described above - cloning screens - I do not envision this being supported by the query parameters, this would be done through the API only due to its complexity.

**Comment by zepumph** [comment]
>For the PhET-iO functionality described above - cloning screens - I do not envision this being supported by the query parameters, this would be done through the API only due to its complexity.

Agreed! I would go one step further and say that I don't think these query parameters should be supported with the phet-io strategy to clone/modify/shuffle screens for PhET-iO. If you are manipulating screens through the phet-io interface, then do that. If using query parameters, do that. I see them as mutually exclusive operations to manipulate screens.

**Comment by samreid** [comment]
Query parameters have the advantage that they are immediately available on startup (as opposed to PhET-iO calls which are asynchronous and designed to be called any time (though can be applied as initial `expressions`)), and easy to implement and test.  I agree that general sim clients wouldn't want to use this query parameter, but I don't think we should rule out query parameters as the implementation strategy for screen cloning.

For instance, we could design a query parameter like: 
`?screenMap=1@introScreen,1@advancedScreen,3@frictionScreen` which would launch the sim with screen 1 named as "introScreen" then another copy of screen 1 named as "advanced screen" then screen 3 named as "frictionScreen".  This would create a new list of screens which can be selected/combined with the previously proposed query parameters.

**Comment by samreid** [comment]
To implement the proposal above, sims mains would need to be retrofitted like so (beers law lab for example):

```js
  var screenFactories = [
    function( tandemName ) {
      return new ConcentrationScreen( tandem.createTandem( tandemName || 'concentrationScreen' ) )
    }, function( tandemName ) {
      return new BeersLawScreen( tandem.createTandem( tandemName || 'beersLawScreen' ) )
    }
  ];

  var screens = Sim.mapScreens( screenFactories );

  SimLauncher.launch( function() {
    var sim = new Sim( beersLawLabTitleString, screens, simOptions );
    sim.start();
  } );
```

I used the optional tandem so the default tandems can be associated with their screens (instead of letting them get separated).

Here's the auxiliary function in Sim.js

```js
    mapScreens: function( screenFactories ) {
      var screens = [];
      var x = phet.chipper.queryParameters.screenMap;
      if ( x.length > 0 ) {
        
        // remap screens like ?screenMap=1@helloScreen,1@thereScreen
        var screenSpecs = x.split( ',' );
        for ( var i = 0; i < screenSpecs.length; i++ ) {
          var screenSpec = screenSpecs[ i ];
          var specStrings = screenSpec.split( '@' );
          var screenIndex = parseInt( specStrings[ 0 ], 10 ) - 1;
          var tandemName = specStrings[ 1 ];
          var screenFactory = screenFactories[ screenIndex ];
          screens.push( screenFactory( tandemName ) );
        }
      }
      else {
        
        // default screens
        for ( var k = 0; k < screenFactories.length; k++ ) {
          var factory = screenFactories[ k ];
          screens.push( factory() );
        }
      }
      return screens;
    }
```

And the query parameter would be like:

```js
    screenMap: {
      type: 'string',
      defaultValue: ''
    },
```

This implementation has the desired behavior:

![image](https://user-images.githubusercontent.com/679486/32252645-e8f6e7b2-be5b-11e7-894f-fa746755fcb6.png)

Note the "helloScreen" and "thereScreen" in the tandems.

**Comment by zepumph** [comment]
It does seem like query parameters are pretty straightforward in that case. What about other data we want to pass for the custom screens like the display name (needs to be URI encoded?)  or the icon. I like this conversation for prototyping, couldn't we have a pretty similar interface but through TPhETIO, that way we could get everything set into the frame before the sim is even launched.

Also should we move back to https://github.com/phetsims/phet-io/issues/1247 for this?  Maybe the best solution for this issue (about home screen query parameter) is to see where phetsims/phet-io#1247 goes and then take it from there.

**Comment by samreid** [comment]
> What about other data we want to pass for the custom screens like the display name (needs to be URI encoded?) or the icon.

I think it makes sense to use Query Parameters to set up the screens (which has to be done before the simulation starts), then rely on the existing PhET-iO APIs for setting titles/icons and further customization.

> couldn't we have a pretty similar interface but through TPhETIO, that way we could get everything set into the frame before the sim is even launched.

Sure, that could work nicely too (just harder to test because it requires a wrapper).  But I think we should try it.

> Also should we move back to phetsims/phet-io#1247 for this?

Yes, and I think we can continue here as long as we perceive the list of sims as a preprocessed list from query parameters or phet-io API.

**Comment by zepumph** [comment]
>... I think we can continue here as long as we perceive the list of sims as a preprocessed list from query parameters or phet-io API.

Agreed, I still recommend the same suggestion from https://github.com/phetsims/joist/issues/451#issuecomment-340081541. 

Still to decide is how they co-exist. Erroring out or ignoring through a heirarchy?

**Comment by kathy-phet** [comment]
For the advanced case of cloning a screen in PhET-iO, I still feel like we can require clients to use the API (using expressions to get it prior to start up?) approach for that. It's an advance use-case and so it doesn't seem like we should worry about presenting query option here -- it seems to overcomplicate the options (to me).

**Comment by zepumph** [comment]
@kathy-phet I responded in https://github.com/phetsims/phet-io/issues/1247#issuecomment-340936354

**Comment by samreid** [comment]
Let's try @zepumph idea to set it through TPhETIO API in https://github.com/phetsims/joist/issues/451#issuecomment-340933680

**Comment by pixelzoom** [comment]
Lots of other questions related to "cloning" screens (regardless of how it's done):
If the same screen appears multiple times, who do you propose to avoid tandem id collisions?
Will multiple copies of the same screen need to be configured in different ways?
What other problems might this cause?
Do we really need to do this?...

**Comment by pixelzoom** [comment]
My input on https://github.com/phetsims/joist/issues/451#issuecomment-340930916...  More complexity, more things to break, more difficult to develop sims.   I'm not digging this at all.

**Comment by samreid** [comment]
> If the same screen appears multiple times, how do you propose to avoid tandem id collisions?

A unique tandem can be given to each clone.

> Will multiple copies of the same screen need to be configured in different ways?

Yes, cloning a screen without changing anything seems to provide no advantage.  The clones can be customized with PhET-iO API calls.

> What other problems might this cause?

It will transform a relatively static API into a dynamic API, which is unconventional and "tricky".

> Do we really need to do this?...

That's a question for @kathy-phet but I think it won't take too long to create a rough draft.  (The main costs will be in maintenance and porting sims to use screen factories instead of instances).

**Comment by pixelzoom** [comment]
@samreid said:
> The main costs will be in maintenance and porting sims to use screen factories instead of instances

So yet another cost for PhET-iO, converting something relatively simple to something complex, "unconventional and tricky".  In https://github.com/phetsims/phet-io/issues/1137#issuecomment-307514863 I identified one of the big management questions as:
> - [ ] How much of an increase in sim cost is acceptable? How much is expected?

@ariel-phet @kathy-phet Has anyone asked this question or answered it? Imo the incremental cost of PhET-iO is very high, perhaps dangerously high.

**Comment by samreid** [comment]
I'm going to commit a proposed strategy for implementing this, we can use it for more concrete discussion and delete it if we wish.  Please see https://github.com/phetsims/phet-io/issues/1247 for details.

**Comment by pixelzoom** [comment]
Since you used BLL as an example... I'd appreciate it if you kept this out of BLL master.

**Comment by samreid** [comment]
Given the other commits in https://github.com/phetsims/phet-io/issues/1247 here's the BLL main that will make it work:

```js
  var screenFactories = [
    function( tandemName ) {
      return new ConcentrationScreen( tandem.createTandem( tandemName || 'concentrationScreen' ) );
    }, function( tandemName ) {
      return new BeersLawScreen( tandem.createTandem( tandemName || 'beersLawScreen' ) );
    }
  ];

  SimLauncher.launch( function() {
    var screens = Sim.createScreens( screenFactories );
    var sim = new Sim( beersLawLabTitleString, screens, simOptions );
    sim.start();
  } );
```

You can test it by launching http://localhost/phet-io-wrappers/create-screens/create-screens.html?sim=beers-law-lab

**Comment by samreid** [comment]
@pixelzoom requested that these changes remain out of bll/master for now. We should review the proposed changes and decide if we are going to support this feature and if this is the way we want to proceed.

**Comment by samreid** [comment]
My main reason for pursuing https://github.com/phetsims/phet-io/issues/1247 now was because it had the potential to impact how we proceed with the proposal in https://github.com/phetsims/joist/issues/451#issuecomment-340848683.  I chose a preprocessing strategy and https://github.com/phetsims/joist/issues/451#issuecomment-340848683 will still work properly on the cloned screens.

I'll investigate https://github.com/phetsims/joist/issues/451#issuecomment-340848683 next.

**Comment by zepumph** [comment]
In https://github.com/phetsims/phet-io/issues/1247#issuecomment-341171937 @kathy-phet said we will not support screen replication, too many red flags pointed out by @pixelzoom in https://github.com/phetsims/phet-io/issues/1247#issuecomment-341138950.

**Comment by pixelzoom** [comment]
11/2/2017 dev meeting:
As proposed in https://github.com/phetsims/joist/issues/451#issuecomment-340873770, we will make these query parameters error out for single-screen sims.

**Comment by samreid** [comment]
I added the `homeScreen` query parameter, and it throws an assertion error when used on a single screen sim.

**Comment by samreid** [comment]
I added the `initialScreen` query parameter, here are some tests to check the behavior:

Note that the screens for FAMB are: net force, motion, friction, acceleration
http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&initialScreen=4&screens=2,4&homeScreen=true
=> home screen + motion | **acceleration** 

http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&initialScreen=4&screens=2,4&homeScreen=false
=> no home screen + motion | **acceleration** 

http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&initialScreen=3
=> home screen available + net force | motion | **friction** | acceleration

http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&initialScreen=3&homeScreen=false&screens=1,2
=> error

http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&initialScreen=3&homeScreen=false&screens=1,1
=> error

http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&homeScreen=false&screens=1,4&initialScreen=0
=> we must allow initialScreen=0 when homeScreen=false (since initialScreen defaults to 0), so leave this as no error and just start on lowest initial screen.

http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&homeScreen=false&screens=1,4&initialScreen=4
=> home screen is available but starts on acceleration screen

It would be nice if we could automatically test these behaviors, but I don't see a simple way to get there.  @phet-steele could you run the preceding tests and recommend some other combinations of query parameters that should also be tested?

**Comment by pixelzoom** [comment]
In Slack, @phet-steele said:
> While I’m testing https://github.com/phetsims/joist/issues/451, are we expecting that ?showHomeScreen and ?screenIndex will NOT do anything? Looks like there was talk about deleting those two, but I still see them in initialize-globals.js

I confirmed that `showHomeScreen` and `screenIndex` are still in initialize-globals.js.  Searching for "queryParameters.showHomeScreen" and "queryParameters.screenIndex",  I see no occurrences, so those query parameters are apparently no longer used.

**Comment by samreid** [comment]
Thanks, I'll remove them now.

**Comment by samreid** [comment]
It looks like navbar items aren't centered when the home button is hidden:
![image](https://user-images.githubusercontent.com/679486/32452728-9d4dca12-c2d7-11e7-9b50-08bdbea768ef.png)

**Comment by samreid** [comment]
I removed screenIndex and showHomeScreen and confirmed that http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&initialScreen=4&screens=2,4&homeScreen=false still works as expected.

**Comment by pixelzoom** [comment]
@samreid said:
> It looks like navbar items aren't centered when the home button is hidden:

Is that are new problem, or did it exist before the changes made for this issue?

**Comment by samreid** [comment]
With the home button, things are (more or lesss?) centered:

![image](https://user-images.githubusercontent.com/679486/32452979-4ec65336-c2d8-11e7-92d5-2fd684eea7fc.png)

This is the first time we are experimenting with removing the home button.  I think it was removed with `setVisible(false)` rather than a strategy that updates the layout.

**Comment by phet-steele** [comment]
> Is that are new problem, or did it exist before the changes made for this issue?

Looks like it happened when we moved the home screen button (so probably unrelated to this issue). The home screen button maybe used to not be considered when centering screen icons?

Production:
<img width="643" alt="screen shot 2017-11-06 at 9 52 55 am" src="https://user-images.githubusercontent.com/5863899/32453031-718e635e-c2d8-11e7-9d2f-db686d744b6c.png">

2.3.0-rc.3:
<img width="605" alt="screen shot 2017-11-06 at 9 53 05 am" src="https://user-images.githubusercontent.com/5863899/32453032-71a0434e-c2d8-11e7-920f-9f16bf9202dc.png">

**Comment by samreid** [comment]
Thanks, I'll create a new issue for centering things and I'll note that centering should still work with or without the home button.

**Comment by phet-steele** [comment]
@samreid these are working great. You have a pretty comprehensive list of tests in https://github.com/phetsims/joist/issues/451#issuecomment-341620806.

One question, did you intend for your last link to have `homeScreen=true`? Otherwise, this link certainly does not have the home screen available.

> http://localhost/forces-and-motion-basics/forces-and-motion-basics_en.html?brand=phet&ea&homeScreen=false&screens=1,4&initialScreen=4
> => home screen is available but starts on acceleration screen
----
> After we confirm the behavior is as-desired and have a set of tests to run (even if manual), then we can reach out to @pixelzoom for review.

Looks like I should hand this off to @pixelzoom.

**Comment by zepumph** [comment]
Reminder to myself: When this is done, it should be updated in https://github.com/phetsims/joist/issues/445#issuecomment-341248255

**Comment by samreid** [comment]
> One question, did you intend for your last link to have `homeScreen=true`?

I think I meant to say "home screen is not available", thanks @phet-steele

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/query-string-machine/issues/30#issuecomment-341507927, we agree to note query parameters that are intended for external use.  That has not been done for `initialScreen` and `homeScreen`.  I'll take care of it now.

**Comment by zepumph** [comment]
I'm not sure if we want to go down this road, but it may be nice to have a jsDoc `@` block annotation that defines this. There is an `@external` but it has little different use. If we go down this road could we adapt it?

**Comment by pixelzoom** [comment]
With `?screens=1,3&initialScreen=2` the assertion failure has this message:

> Assertion failed: screen not found

The message should identify the screen index. I'll take care of this.

**Comment by pixelzoom** [comment]
This combination seems odd: `?homeScreen=false&screens=1,3&initialScreen=0`.   We've explicitly asked for the home screen to be the initial screen, but there is no home screen.   This succeeds because `initialScreen` is ignored if its value is `0` (see Sim.js line 127), which seems a bit brittle.  And the default value of `initialScreen` is `0`, so we can't detect that it was explicitly set to '0'.  I'm OK with this, but wanted to point it out in case anyone else thinks it should be handled.

**Comment by samreid** [comment]
I added an assertion error that fails for the combination of homeScreen=false and initialScreen=0, @pixelzoom can you please review?

**Comment by pixelzoom** [comment]
> I added an assertion error ...

Hmmm... That reminds me. Assertions are not appropriate for checking any of these query parameters.  They should throw and Error regardless of whether assertions are enabled.

**Comment by pixelzoom** [comment]
To clarify... `assert` is fine for checking query parameters that are for internal use.  But for query parameters available for external, use `throw Error`.

**Comment by pixelzoom** [comment]
> I added an assertion error ...

Hmmm... That reminds me. Assertions are not appropriate for checking any of these query parameters.  They should `throw Error` regardless of whether assertions are enabled.

In general... `assert` is fine for checking query parameters that are for internal use. But for query parameters available for external, use `throw Error`.

**Comment by samreid** [comment]
I asked on slack:

Is our idiom for errors like so?

 ```if (problematic) throw new Error('…');```
or should I use `assert(!probematic,'error message')` without the `&&`?
or create some sort of new error function for use in cases like this?

**Comment by pixelzoom** [comment]
@samreid  In your post to Slack, you didn't provide information about the nature of "problematic". And that determines whether to use assert of Error.  Please describe your understanding of when it's appropriate to use assertions vs throwing an Error.




If you use `assert` without `assert &&`, it will fail.

**Comment by pixelzoom** [comment]
From http://www.suodenjoki.dk/us/archive/2011/assert.htm:

> “an assertion is documentation that reason about how the design and the code should work. They are pre-conditions, post-conditions and loop invariants that the programmer make up to document expectations for a programs execution. It is the programmers contract with his design understanding.” … “Don’t use assert for error handling. It is not intended for this. It has nothing to do with error handling or exception handling at all.”

**Comment by samreid** [comment]
In the preceding commit I converted the Sim.js query string checks from assertions to errors.  @pixelzoom can you please review?

**Comment by pixelzoom** [comment]
Looks good.  I fixed one of the error messages, which was referring to the old `screenIndex` query param.  

Closing.
